File: Java/프로그래머스/2/60058. 괄호 변환/괄호 변환.java
import java.util.ArrayDeque;

class Solution {
    public String solution(String p) {
        return program(p);
    }
    
    private String program(String w) {
        if (w.isEmpty()) return w;
        String[] dvided = divide(w);
        String u = dvided[0];
        String v = dvided[1];
        if (validateString(u)) {
            return u + program(v);
        } else {
            StringBuilder sb = new StringBuilder();
            sb.append('(');
            sb.append(program(v));
            sb.append(')');
            for (int index = 1; index < u.length() - 1; index++) {
                sb.append(u.charAt(index) == '(' ? ')' : '(');
            }
            
            return sb.toString();
        }
    }
    
    private String[] divide(String w) {
        boolean isUCorrect = true;
        String u = "";
        String v = "";
        
        int openCount = 0;
        int closeCount = 0;
        int subStringIndex = w.length() - 1;
        for (int i = 0; i < w.length(); i++) {
            char c = w.charAt(i);
            if (c == '(') {
                openCount++;
            } else {
                closeCount++;
            }
            if (openCount == closeCount) {
                subStringIndex = i;
                break;
            }
        }
        u = w.substring(0, subStringIndex + 1);
        if (subStringIndex + 1 < w.length()) {
            v = w.substring(subStringIndex + 1, w.length());
        }
        
        return new String[]{u, v};
    }
    
    private boolean validateString(String s) {
        ArrayDeque<Character> ad = new ArrayDeque<>();
        
        for (char c : s.toCharArray()) {
            if (c == '(') {
                ad.add(c);
            } else {
                if (ad.isEmpty()) {
                    return false;
                } else {
                    ad.pollLast();
                }
            }
        }
        
        return true;
    }
}

File: Swift/프로그래머스/2/60058. 괄호 변환/괄호 변환.swift
import Foundation

func solution(_ p:String) -> String {
    return program(p)
}

func program(_ w: String) -> String {
    if w.isEmpty {
        return w
    }
    let divided = divide(w)
    let u = divided.0
    let v = divided.1
    if validate(u) {
        return u + program(v)
    } else {
        var result = "("
        result.append(program(v))
        result.append(")")
        for i in 1..<u.count-1 {
            result.append(u[u.index(u.startIndex, offsetBy: i)] == "(" ? ")" : "(")
        }
        return result
    }
}

func divide(_ w: String) -> (String, String) {
    var balance = 0
    var index: String.Index = w.startIndex
    var u: String = ""
    var v: String = ""
    
    for i in 0..<w.count {
        index = w.index(w.startIndex, offsetBy: i)
        let c = w[index]
        
        u.append(c)
        
        if c == "(" {
            balance += 1
        } else {
            balance -= 1
        }
        
        if balance == 0 {
            break;
        }
    }
    
    if index < w.endIndex {
        let vIndex = w.index(index, offsetBy: 1)
        v = String(w[vIndex..<w.endIndex])
    }
    
    return (u, v)
}

func validate(_ w: String) -> Bool {
    var balance = 0
    for i in 0..<w.count {
        let index = w.index(w.startIndex, offsetBy: i)
        if w[index] == "(" {
            balance += 1
        } else {
            balance -= 1
        }
        
        if (balance < 0) {
            return false
        }
    }
    
    return balance == 0
}